extensions/socket/internal.m

```diff
@@ -0,0 +1,362 @@
+    }
+}
+
+
+@implementation HSAsyncSocket
+
+- (id)init {
+    return [super initWithDelegate:self delegateQueue:dispatch_get_main_queue()];
+}
+
+- (void)socket:(HSAsyncSocket *)sock didConnectToHost:(NSString *)host port:(uint16_t)port {
+    [[LuaSkin shared] logInfo:@"Socket connected"];
+}
+
+- (void)socketDidDisconnect:(HSAsyncSocket *)sock withError:(NSError *)err {
+    [[LuaSkin shared] logInfo:@"Socket disconnected"];
+}
+
+- (void)socket:(HSAsyncSocket *)sock didWriteDataWithTag:(long)tag {
+    [[LuaSkin shared] logInfo:@"Data written to socket"];
+}
+
+- (void)socket:(HSAsyncSocket *)sock didReadData:(NSData *)data withTag:(long)tag  {
+    [[LuaSkin shared] logInfo:@"Data read from socket"];
+
+    callback(self, data);
+}
+
+@end
+
+
+/// hs.socket.new(host, port) -> hs.socket object
+/// Constructor
+/// Creates an asynchronous TCP socket object for reading (with callbacks) and writing
+///
+/// Parameters:
+///  * host - A string containing the hostname or IP address
+///  * port - A port number [1024-65535]. Ports [1-1023] are privileged
+///  * fn - An optional callback function accepting a single parameter to process data. Can be set with `:setCallback`
+///
+/// Returns:
+///  * An `hs.socket` object
+///
+static int socket_new(lua_State *L) {
+    LuaSkin *skin = [LuaSkin shared];
+    [skin checkArgs:LS_TSTRING, LS_TNUMBER, LS_TFUNCTION|LS_TNIL|LS_TOPTIONAL, LS_TBREAK];
+
+    HSAsyncSocket *asyncSocket = [[HSAsyncSocket alloc] init];
+
+    NSString *theHost = [skin toNSObjectAtIndex:1];
+    NSNumber *thePort = [skin toNSObjectAtIndex:2];
+    if (lua_type(L, 3) == LUA_TFUNCTION) {
+        lua_pushvalue(L, 3);
+        asyncSocket.callback = [skin luaRef:refTable];
+    }
+
+    NSError *err;
+    if (![asyncSocket connectToHost:theHost onPort:[thePort unsignedShortValue] error:&err]) {
+        [skin logError:[NSString stringWithFormat:@"Unable to connect: %@", err]];
+    }
+
+    [skin pushNSObject:asyncSocket];
+    return 1;
+}
+
+/// hs.socket.read(delimiter) -> self
+/// Method
+/// Read data from the socket. Data is passed to the callback function
+///
+/// Parameters:
+///  * delimiter - Either a number of bytes to read, or a string delimiter such as `\n` or `\r\n`. Data is read up to and including the delimiter
+///
+/// Returns:
+///  * The `hs.socket` object
+///
+static int socket_read(lua_State *L) {
+    LuaSkin *skin = [LuaSkin shared];
+    [skin checkArgs:LS_TUSERDATA, USERDATA_TAG, LS_TSTRING|LS_TNUMBER, LS_TBREAK];
+
+    HSAsyncSocket*  asyncSocket = [skin luaObjectAtIndex:1 toClass:"HSAsyncSocket"];
+
+    if (!asyncSocket.callback) {
+        [skin logWarn:@"No callback! Defaulting to print()"];
+    }
+
+    switch (lua_type(L, 2)) {
+        case LUA_TNUMBER: {
+            NSNumber *bytesToRead = [skin toNSObjectAtIndex:2];
+            NSUInteger bytes = [bytesToRead unsignedIntegerValue];
+            [asyncSocket readDataToLength:bytes withTimeout:-1 tag:-1];
```

Comment:
I wonder if we might want to make the timeout a property on the socket object, and expose a method to read/set it?